OpenRA.Mods.Common/Activities/Resupply.cs

```diff
@@ -114,36 +166,42 @@ public override Activity Tick(Actor self)
 				if (!self.Owner.PlayerActor.Trait<PlayerResources>().TakeCash(cost, true))
 				{
 					remainingTicks = 1;
-					return this;
+					return;
 				}
 
 				self.InflictDamage(host.Actor, new Damage(-hpToRepair));
-
-				foreach (var depot in host.Actor.TraitsImplementing<INotifyRepair>())
-					depot.RepairTick(host.Actor, self);
-
 				remainingTicks = repairsUnits.Info.Interval;
 			}
 			else
 				--remainingTicks;
-
-			return this;
 		}
 
-		protected override void OnLastRun(Actor self)
+		void RearmTick(Actor self)
 		{
-			if (host.Actor.IsDead)
-				return;
+			rearmComplete = true;
+			foreach (var pool in rearmable.RearmableAmmoPools)
+			{
+				if (!pool.FullAmmo())
+				{
+					Reload(self, host.Actor, pool);
+					rearmComplete = false;
+				}
+			}
 
-			foreach (var depot in host.Actor.TraitsImplementing<INotifyRepair>())
-				depot.AfterRepair(host.Actor, self);
+			if (rearmComplete)
+				activeResupplyTypes &= ~ResupplyType.Rearm;
 		}
 
-		protected override void OnActorDispose(Actor self)
+		void Reload(Actor self, Actor host, AmmoPool ammoPool)
```

Comment:
This is only used from one place and could be inlined.

Reply:
Done.